solr/core/src/java/org/apache/solr/core/CachingDirectoryFactory.java

```diff
@@ -339,10 +332,10 @@ public boolean exists(String path) throws IOException {
    * (non-Javadoc)
    *
    * @see org.apache.solr.core.DirectoryFactory#get(java.lang.String,
-   * java.lang.String, boolean)
+   * java.lang.String, boolean, java.util.function.Function)
    */
   @Override
-  public final Directory get(String path, DirContext dirContext, String rawLockType)
+  public final Directory get(String path, DirContext dirContext, String rawLockType, Function<Directory, Directory> wrappingFunction)
```

Comment:
is cachingdirectoryfactory supposed to extend delegatingDF now? I'm confused about the relationship here.

Reply:
I don't think CachingDF should extend DelegatingDF. Here the goal with this new param is to allow the caller to have a hook when the Directory is created internally. Otherwise there is no way a DelegatingDF "A" can delegate to any CachingDF "B" while leveraging the cache inside B and at the same time have control over the class of Directory cached and returned.
Another option could be to first extend the CachingDF with a B' class that overrides the createDirectory() method. But that means a DelegatingDF can only delegate to a specific class B' extending CachingDF. But this is too limiting.

Reply:
A sample usage could be a BlobDF extending DelegatingDF to delegate to *any* configured DF, probably MMapDF which is a CachingDF. BlobDF would need to create BlobDir that delegates to a MMapDir. BlobDir cannot wrap the MMapDir outside of the CachingDF, because for the cache to actually work, it has to cache the BlobDir instance. That's why the idea is to provide a wrapper function to wrap a MMapDir with a BlobDir when it is created internally by the CachingDF.